ring/spread_minimizing_token_generator.go

```diff
@@ -303,3 +337,27 @@ func (t *SpreadMinimizingTokenGenerator) generateTokensByInstanceID() map[int]To
 
 	return tokensByInstanceID
 }
+
+func (t *SpreadMinimizingTokenGenerator) CanJoin(instances map[string]InstanceDesc) error {
```

Comment:
I can't understand how this could work. The input `instances` is just a one time snapshot of the ring. You need to move the `for` loop outside (back into `Lifecycler`) and call it each time with the updated map of `instances`.

You also need to handle the `ctx` cancellation in the for loop and you need some throttling. In the for loop below you're looping with no throttling, while we should check the ring every X time.

In dskit there's `backoff`. Use it: it will handle throttling, backoff and context cancellation for you.